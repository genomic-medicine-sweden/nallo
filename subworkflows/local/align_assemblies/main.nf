include { MINIMAP2_ALIGN   } from '../../../modules/nf-core/minimap2/align/main'
include { MINIMAP2_INDEX   } from '../../../modules/nf-core/minimap2/index/main'
include { SAMTOOLS_MERGE   } from '../../../modules/nf-core/samtools/merge/main'
include { SAMTOOLS_VIEW    } from '../../../modules/nf-core/samtools/view/main'
include { SAMTOOLS_CONVERT } from '../../../modules/nf-core/samtools/convert/main'
include { TAGBAM           } from '../../../modules/nf-core/tagbam/main'

workflow ALIGN_ASSEMBLIES {

    take:
    ch_assembly // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fasta    // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai      // channel: [mandatory] [ val(meta), path(fai)   ]
    cram_output // bool: Publish alignments as CRAM (true) or BAM (false)

    main:
    ch_versions = Channel.empty()

    MINIMAP2_INDEX (
        ch_fasta
    )
    ch_versions = ch_versions.mix(MINIMAP2_INDEX.out.versions)

    MINIMAP2_ALIGN (
        ch_assembly,
        MINIMAP2_INDEX.out.index.collect(),
        true,
        'bai',
        false,
        false
    )
    ch_versions = ch_versions.mix(MINIMAP2_ALIGN.out.versions)

    SAMTOOLS_VIEW (
        MINIMAP2_ALIGN.out.bam.join(MINIMAP2_ALIGN.out.index, failOnMismatch:true, failOnDuplicate:true),
        [[],[]],
        []
    )
    ch_versions = ch_versions.mix(SAMTOOLS_VIEW.out.versions)

    TAGBAM (
        SAMTOOLS_VIEW.out.bam
    )
    ch_versions = ch_versions.mix(TAGBAM.out.versions)

    TAGBAM.out.bam
        .map { meta, bam -> [ meta - meta.subMap('haplotype'), bam ] }
        .groupTuple(size: 2)
        .set { ch_assemblies_per_sample }

    SAMTOOLS_MERGE (
        ch_assemblies_per_sample,
        [[],[]],
        [[],[]]
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MERGE.out.versions)

    // Publish alignment as CRAM if requested
    if (cram_output) {
        SAMTOOLS_CONVERT(
            SAMTOOLS_MERGE.out.bam.join(SAMTOOLS_MERGE.out.bai, failOnDuplicate: true, failOnMismatch: true),
            ch_fasta,
            ch_fai
        )
        ch_versions = ch_versions.mix(SAMTOOLS_CONVERT.out.versions)
    }

    emit:
    bam      = SAMTOOLS_MERGE.out.bam // channel: [ val(meta), path(bam) ]
    bai      = SAMTOOLS_MERGE.out.bai // channel: [ val(meta), path(bai) ]
    versions = ch_versions            // channel: [ versions.yml ]
}

