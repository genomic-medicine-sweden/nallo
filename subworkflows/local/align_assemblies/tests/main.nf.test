nextflow_workflow {

    name "Test Workflow ALIGN_ASSEMBLIES"
    script "../main.nf"
    workflow "ALIGN_ASSEMBLIES"
    config "./nextflow.config"

    setup {
        run("HIFIASM") {
            script "../../../../modules/nf-core/hifiasm/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'hg38' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/pacbio/fastq/test_hifi.fastq.gz', checkIfExists: true),
                    []
                ])
                input[1] = [[],[],[]]
                input[2] = [[],[],[]]
                """
            }
        }

        run("GFASTATS") {
            script "../../../../modules/nf-core/gfastats/main.nf"
            process {
                """
                input[0] = HIFIASM.out.paternal_contigs
                    .map { meta, fasta -> [ meta + [ 'haplotype': 1 ], fasta ] }
                    .mix(
                        HIFIASM.out.maternal_contigs
                        .map { meta, fasta -> [ meta + [ 'haplotype': 2 ], fasta ] }
                    )
                input[1] = 'fasta'
                input[2] = ''
                input[3] = ''
                input[4] = [[],[]]
                input[5] = [[],[]]
                input[6] = [[],[]]
                input[7] = [[],[]]
                """
            }
        }

    }

    test("bam") {

        when {
            workflow {
                """
                input[0] = GFASTATS.out.assembly
                input[1] = Channel.of([
                    [ id: 'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    bam(workflow.out.bam.get(0).get(1)).getReadsMD5(),
                    file(workflow.out.bai.get(0).get(1)).name
                ).match() }
            )
        }
    }

    test("bam - stub") {
        options "-stub"
        when {
            workflow {
                """
                input[0] = GFASTATS.out.assembly
                input[1] = Channel.of([
                    [ id: 'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

    }
}
