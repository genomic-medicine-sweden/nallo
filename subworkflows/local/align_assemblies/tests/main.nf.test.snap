{
    "bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "hg38"
                        },
                        "hg38.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "hg38"
                        },
                        "hg38.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                    "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                    "versions.yml:md5,737f29f0cdcb2c85965487e46110fc4b",
                    "versions.yml:md5,8065259e8f1e0e9af6f2b56effbe0156",
                    "versions.yml:md5,8065259e8f1e0e9af6f2b56effbe0156",
                    "versions.yml:md5,82e6ac9ca240d79971cb197be4b30647",
                    "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0",
                    "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0"
                ],
                "bai": [
                    [
                        {
                            "id": "hg38"
                        },
                        "hg38.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "hg38"
                        },
                        "hg38.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                    "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                    "versions.yml:md5,737f29f0cdcb2c85965487e46110fc4b",
                    "versions.yml:md5,8065259e8f1e0e9af6f2b56effbe0156",
                    "versions.yml:md5,8065259e8f1e0e9af6f2b56effbe0156",
                    "versions.yml:md5,82e6ac9ca240d79971cb197be4b30647",
                    "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0",
                    "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-02-17T14:36:24.630337248"
    },
    "bam": {
        "content": [
            [
                "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                "versions.yml:md5,493d8cf85cee34a20902289619bc5d5b",
                "versions.yml:md5,737f29f0cdcb2c85965487e46110fc4b",
                "versions.yml:md5,82e6ac9ca240d79971cb197be4b30647",
                "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0",
                "versions.yml:md5,b1e934cf9c14105eed4afe0744b1fcd0",
                "versions.yml:md5,dd66d91ce97e7998f21215a8dfd1fa60",
                "versions.yml:md5,dd66d91ce97e7998f21215a8dfd1fa60"
            ],
            "43bb6ef7e9eccece743204535a3abbd6",
            "hg38.bam.bai"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-02-17T14:35:36.622658596"
    }
}